### PR TITLE
[improve][pip] Change cursor`s properties to store chunk ID map.

### PR DESCRIPTION
**Pull request by liangyepianzhou**
### Motivation
Chunk messages can be effectively filtered on the broker side. Ensure that chunk messages work normally after enabling deduplication and the topic has no duplicate chunks.

### Modifications
1. Add `chunkIDPushed` and `chunkIDPersisted` to store the chunk of each producer`s ongoing chunk messages. It will be used to check whether the chunks in a single message are duplicated.
2. Optimize the `properties` of the `MarkDeleteEntry` from `Map<String, Long>` to `Map<String, String>`.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [ ] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository:

### CODE REVIEW

File: pip/pip-295.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        }
+    }
+
+    recoveredCursor(recoveredPosition, recoveredProperties, recoveredCursorProperties, null);
+```
+Change to:
+```java
+    // Recover properties map
+    Map<String, String> recoveredProperties;
+    if (info.getPropertiesCount() == 0 && info.getmarkDeletePropertiesCount() == 0) {
```
Comment (poorbarcode) [inline pip/pip-295.md:102]: The `info` means `ManagedCursorInfo`, right?

File: pip/pip-295.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+# Motivation
+
+In the earliest design, all the chunks in a single chunk message have the same sequence ID which causes the chunk message can not work when enabling deduplication. For example, we have a  chunk message consisting of chunk-1 and chunk-2. When Broker receives chunk-1, it will update the last sequence ID to the sequence ID of chunk-1. And then, when the broker gets chunk-2, the chunk-2 will be dropped by depublication.
+I opened a [PR](https://github.com/apache/pulsar/pull/20948) to resolve this case. It allowed the chunks of a single chunk message to use the same sequence ID and filter duplicated chunks in a single-chunk message on the consumer side.
+It can resolve message duplication end to end, but the message duplication still exists in the topic.
+
+# Goals
+
+## In Scope
+Chunk messages can be effectively filtered on the broker side. Ensure that chunk messages work normally after enabling deduplication and the topic has no duplicate chunks.
```
Comment (poorbarcode) [inline pip/pip-295.md:17]: **Background:***: There are two properties in the metadata of the cursor
- `properties<String, Long>`: used to maintain the last sequence of producer-sent messages<sup>[1]</sup>.
  - PIP: https://github.com/apache/pulsar/wiki/PIP-6:-Guaranteed-Message-Deduplication
  - PR: https://github.com/apache/pulsar/pull/744
-  `cursorProperties<String, String>`: used to maintain the subscription properties. 
  - PIP: https://github.com/apache/pulsar/issues/12269 
  - PR: https://github.com/apache/pulsar/issues/15750

**[1]**: a structure of `properties`:
```yaml
properties: 
  - "producer_name_1" : {{last_persist_sequence_1}}
  - "producer_name_2" : {{last_persist_sequence_2}}
```

----

In this PIP, you want to change `properties<String, Long>` to `properties<String, Long>`, right? Could you also explain this change here?

File: pip/pip-295.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+    @VisibleForTesting
+    final ConcurrentOpenHashMap<String, Integer> chunkIDPersisted =
+            ConcurrentOpenHashMap.<String, Integer>newBuilder()
+            .expectedItems(16)
+            .concurrencyLevel(1)
+            .build();
+```
+
+Optimize the `properties` of the `MarkDeleteEntry` from `Map<String, Long>` to `Map<String, String>`. In the depublication design, the ' MarkDeleteEntry' properties are used as a snapshot to store the sequence ID map. After introducing the chunk ID map, it cannot hold two long for each producer. So we hope to change the `MarkDeleteEntry' properties from `Map<String, Long>` to `Map<String, String>` to make it more flexible.
```
Comment (poorbarcode) [inline pip/pip-295.md:49]: Could you add a demo to describe the structure of the attribute `properties` of cursor metadata that you wanted after this PIP?

File: pip/pip-295.md (hunk 1 above)
Comment (liangyepianzhou) [reply pip/pip-295.md:102]: It's true.

**Pull request closed without merging**